Add diagnostic to flag unsupported usages of initializer APIs

crisbeto opened the pull request:
Adds a diagnostic that will flag any usages of initializer APIs outside of initializers. Includes the following changes:

### refactor(compiler-cli): add API to check if imports exist
Adds a couple of APIs to the `ImportedSymbolsTracker` that allow us to quickly check if a specific symbol is imported in a file.

### refactor(compiler-cli): introduce infrastructure for running additional checks against TypeScript files
Adds the new `SourceFileValidator` that will be used to check for file-level issues that may prevent Angular from working, like invoking the `input()` function outside of an initializer. Currently only one check is planned, but this setup will allow us to easily add more in the future.

### fix(compiler-cli): add diagnostic if initializer API is used outside of an initializer
Adds a rule that will produce a diagnostic when an initializer-based API is used outside of an initializer.

### fix(compiler-cli): report cases where initializer APIs are used in a non-directive class
Expands the check for initializer APIs to also flag when the function is called on a class that isn't a component or directive.

### refactor(compiler-cli): allow source file validator rule to produce a single diagnostic
Allows for `SourceFileValidatorRule.checkNode` to produce a single diagnostic. The most common case should be one diagnostic per node so this allows us to save some array allocations.

Fixes #54381.

crisbeto committed:
refactor(compiler-cli): add API to check if imports exist

Adds a couple of APIs to the `ImportedSymbolsTracker` that allow us to quickly check if a specific symbol is imported in a file.

crisbeto committed:
refactor(compiler-cli): introduce infrastructure for running additional checks against TypeScript files

Adds the new `SourceFileValidator` that will be used to check for file-level issues that may prevent Angular from working, like invoking the `input()` function outside of an initializer. Currently only one check is planned, but this setup will allow us to easily add more in the future.

crisbeto commented on packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:
```diff
@@ -0,0 +1,78 @@
+export class InitializerApiUsageRule implements SourceFileValidatorRule {
+  constructor(
+      private reflector: ReflectionHost, private importedSymbolsTracker: ImportedSymbolsTracker) {}
+
```
It's worth noting that this check currently doesn't capture cases like `input.call(null)`. I wanted to keep it simple for now and see if they even come up.

devversion reviewed (approved):
Nice 🎉

crisbeto commented on packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:
```diff
@@ -0,0 +1,78 @@
+  {functionName: 'viewChildren', owningModule: '@angular/core'},
+  {functionName: 'contentChild', owningModule: '@angular/core'},
+  {functionName: 'contentChildren', owningModule: '@angular/core'},
+];
```
Yeah it's doable, although I guess we'll still have to remember to update this file if we add ned APIs.

devversion replied:
I think we should do this, especially with the changes from the other PR. Can follow-up though

crisbeto commented on packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:
```diff
@@ -0,0 +1,78 @@
+    if (!node.parent || !ts.isCallExpression(node) ||
+        (ts.isPropertyDeclaration(node.parent) && node.parent.initializer === node)) {
+      return null;
+    }
```
Hmm, good question. I was under the impression that the handlers would flag cases like that, but I haven't checked it.

devversion replied:
Should we add it here?

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
Reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

crisbeto committed:
fix(compiler-cli): add diagnostic if initializer API is used outside of an initializer

Adds a rule that will produce a diagnostic when an initializer-based API is used outside of an initializer.

Fixes #54381.

crisbeto committed:
fix(compiler-cli): report cases where initializer APIs are used in a non-directive class

Expands the check for initializer APIs to also flag when the function is called on a class that isn't a component or directive.

crisbeto committed:
refactor(compiler-cli): allow source file validator rule to produce a single diagnostic

Allows for `SourceFileValidatorRule.checkNode` to produce a single diagnostic. The most common case should be one diagnostic per node so this allows us to save some array allocations.

crisbeto commented:
This change should be ready to go now. I've rebased, addressed the feedback and expanded the check to also cover classes that aren't components/directives.

dylhunn commented:
This PR was merged into the repository by commit 336916ceb07947bc78d4483f95d086e7f4841a30.
